rbc/irtools.py

```diff
@@ -331,80 +369,63 @@ def compile_to_LLVM(functions_and_signatures,
-                                          library=main_library,
-                                          locals={},
-                                          pipeline_class=pipeline_class)
-            make_wrapper(fname, args, return_type, cres,
-                         verbose=debug)
-
-    seen = set()
-    for _library in main_library._linking_libraries:
-        if _library not in seen:
-            seen.add(_library)
-            main_module.link_in(
-                _library._get_module_for_linking(), preserve=True,
-            )
+        for fid, sig in signatures.items():
+            fname = compile_instance(func, sig, target, typing_context,
+                                     target_context, pipeline_class,
+                                     main_library,
+                                     debug=debug)
+            if fname is not None:
+                succesful_fids.append(fid)
+                function_names.append(fname)
 
     main_library._optimize_final_module()
 
-    # Catch undefined functions:
-    used_functions = catch_undefined_functions(main_module, target, function_names)
-
-    unused_functions = [f.name for f in main_module.functions
-                        if f.name not in used_functions]
-
-    if debug:
-        print('compile_to_IR: the following functions are used')
-        for fname in used_functions:
-            lf = main_module.get_function(fname)
-            print('  [ALIVE]', fname, 'with', lf.linkage)
-
-    if unused_functions:
-        if debug:
-            print('compile_to_IR: the following functions are not used'
-                  ' and will be removed:')
-        for fname in unused_functions:
-            lf = main_module.get_function(fname)
-            if lf.is_declaration:
-                # if the function is a declaration,
-                # we just put the linkage as external
-                lf.linkage = llvm.Linkage.external
-            else:
-                # but if the function is not a declaration,
-                # we change the linkage to private
-                lf.linkage = llvm.Linkage.private
-            if debug:
-                print('  [DEAD]', fname, 'with', lf.linkage)
+    # Remove unused defined functions and declarations
+    used_symbols = defaultdict(set)
+    for fname in function_names:
+        for k, v in get_called_functions(main_library, fname).items():
+            used_symbols[k].update(v)
+
+    all_symbols = get_called_functions(main_library)
+
+    unused_symbols = defaultdict(set)
+    for k, lst in all_symbols.items():
+        if k == 'libraries':
+            continue
+        for fn in lst:
+            if fn not in used_symbols[k]:
+                unused_symbols[k].add(fn)
+
+    changed = False
+    for f in main_module.functions:
+        fn = f.name
+        if fn.startswith('llvm.'):
+            if f.name in unused_symbols['intrinsics']:
+                f.linkage = llvm.Linkage.external
+                changed = True
```

Comment:
I think here you can leverage `libtools.LLVMIntrinsic.check` or `libtools.NVVMIntrisic.check` to check if the symbol is really an intrinsic.

Reply:
I'll leave it for the follow-up PRs.